api/public/tests/features/steps/device_email.py

```diff
@@ -16,59 +16,79 @@
-    context.email_response = context.client.put(
-        '/v1/device/{uuid}/message'.format(uuid=device_id),
-        data=json.dumps(email_request),
-        content_type='application_json',
-        headers=dict(Authorization='Bearer {token}'.format(token=access_token))
-    )
+    The SendGrid call to send email is mocked out because we trust that the library
+    is coded to correctly send email via sendgrid.  That, and testing sent emails is
+    a real pain in the arse.
+    """
+    with patch("selene.util.email.email.Mail") as message_patch:
+        context.message_patch = message_patch
+        with patch("selene.util.email.email.SendGridAPIClient") as sendgrid_patch:
+            sendgrid_patch.return_value, post_mock = _define_sendgrid_mock()
+            context.sendgrid_patch = sendgrid_patch
+            context.post_mock = post_mock
+            _call_email_endpoint(context)
 
 
-@then('an email should be sent to the user\'s account that owns the device')
-def validate_response(context):
-    response = context.email_response
-    assert_that(response.status_code, equal_to(HTTPStatus.OK))
-    email_client: MagicMock = context.client_config['EMAIL_CLIENT']
-    email_client.send_message.assert_called()
+@when("an unpaired or unauthenticated device attempts to send an email")
+def send_email_invalid_device(context):
+    """Call the email endpoint with an invalid device ID to test authentication."""
+    _call_email_endpoint(context, device_id=str(uuid.uuid4()))
+
 
+def _define_sendgrid_mock():
+    """Go through some insane hoops to setup the correct Mock for the Sendgrid code."""
+    post_return_value = MagicMock(name="post_return")
+    post_return_value.status_code = 200
+    post_mock = MagicMock(name="post_mock")
+    post_mock.return_value = post_return_value
+    send_mock = MagicMock(name="send_mock")
+    send_mock.post = post_mock
+    mail_mock = MagicMock(name="mail_mock")
+    mail_mock.send = send_mock
+    client_mock = MagicMock(name="client_mock")
+    client_mock.mail = mail_mock
+    sendgrid_mock = MagicMock(name="sendgrid_mock")
+    sendgrid_mock.client = client_mock
 
-@when('the email endpoint is called by a not allowed device')
-@patch('smtplib.SMTP')
-def send_email_invalid_device(context, email_client):
-    context.client_config['EMAIL_CLIENT'] = email_client
-    context.email_invalid_response = context.client.put(
-        '/v1/device/{uuid}/email'.format(uuid=str(uuid.uuid4())),
-        data=json.dumps(email_request),
-        content_type='application_json'
+    return sendgrid_mock, post_mock
+
+
+def _call_email_endpoint(context, device_id=None):
+    """Build the request to the email endpoint and issue a call."""
+    if device_id is None:
+        login = context.device_login
+        request_device_id = login["uuid"]
+        request_headers = dict(Authorization=f"Bearer {login['accessToken']}")
+    else:
+        request_device_id = device_id
+        request_headers = dict()
```

Comment:
should this have an invalid Bearer to ensure it is rejecting the call if the `accessToken` doesn't match as opposed to just missing?

Reply:
@chrisveilleux did you see this one?

Reply:
It still fails as designed because the device ID is not valid.  I suppose there is a chance the generated UUID matches, but a very small one.  I can add a dummy bearer token for good measure